### PR TITLE
Add feature #1572 which gives support for multiple candidate sequences

### PR DESCRIPTION
**Pull request by enzoampil**
**Multiple candidate sequences can be generated by setting ```num_samples  >  1``` (still 1 by default).**

EXAMPLE with ```num_samples == 2``` for a GPT2 model:

```
INPUT:
Why did the chicken

OUTPUT:
 cross the road <eoq>  To go to the other side. <eoa>  
 eat food <eoq>  Because it was hungry <eoa>

```
(above is illustrative w some words changed from the actual output)

**UPDATE:** Multiple candidate sequences can now be generated with _repetition penalty_ and ```top_k_top_p_filtering``` applied separately to each candidate. This allows for independent probability distributions across candidate sequences.

~Samples are generated with replacement to allow for sequences that have similar tokens at the same index (e.g. [CLS], stopwords, punctuations).~

~When ```temperature == 0```, the tokens returned are the top ```num_samples``` logits (first sample gets top 1, second gets top 2, and so on). I realize this might not be the best implementation because it doesn't allow for similar tokens at the same index across samples. I will consider later changing this to just returning ```num_samples``` copies of the top 1 logits (argmax).~

### CODE REVIEW

**Comment by enzoampil** [comment]
Please note that main() now returns a list with ```num_samples``` elements inside.

Because of this, the test for run_generation.py should be updated to test for ```length``` for each element within the list. This explains why **build_py3_torch** test failed.

I will update ```test_generation``` for this.

**Comment by enzoampil** [comment]
Latest commit 17dd64e now applies repetition penalty and ```top_k_top_p_filtering``` to each candidate sequence separately.

**Comment by julien-c** [comment]
Thanks @enzoampil 

Superseded by https://github.com/huggingface/transformers/pull/1333 which was just merged to master.

Let me know if this works for your use case.
